test/Libraries/DynamoPythonTests/PythonEditTests.cs

```diff
@@ -511,5 +512,107 @@ public void Python_CanReferenceDynamoServicesExecutionSession()
                 AssertPreviewCount(guid, 2);
             }
         }
+
+        [Test]
+        public void CPythonClassCanBeUsedInDownStreamNode()
+        {
+            // open test graph
+            var examplePath = Path.Combine(TestDirectory, @"core\python", "cpythoncustomclass.dyn");
+            ViewModel.OpenCommand.Execute(examplePath);
+
+            var downstream1 = ViewModel.Model.CurrentWorkspace.Nodes.First(x => x.Name == "downstream1");
+            var downstream2 = ViewModel.Model.CurrentWorkspace.Nodes.First(x => x.Name == "downstream2");
+            
+            ViewModel.HomeSpace.Run();
+            AssertPreviewValue(downstream1.GUID.ToString(), "firstName");
+            AssertPreviewValue(downstream2.GUID.ToString(), "firstNamelastname");
+
+        }
+        [Test]
+        public void CPythonClassCanBeModifiedInDownStreamNode()
+        {
+            // open test graph
+            var examplePath = Path.Combine(TestDirectory, @"core\python", "cpythoncustomclass_modified.dyn");
+            ViewModel.OpenCommand.Execute(examplePath);
+
+            var downstream1 = ViewModel.Model.CurrentWorkspace.Nodes.First(x => x.Name == "downstream1");
+            var downstream2 = ViewModel.Model.CurrentWorkspace.Nodes.First(x => x.Name == "downstream2");
+
+            ViewModel.HomeSpace.Run();
+            AssertPreviewValue(downstream2.GUID.ToString(), "joe");
+        }
+
+        [Test]
+        public void TwoCPythonHandlesReturnedFromSameNodeHaveSameHandleID()
+        {
+            // open test graph
+            var examplePath = Path.Combine(TestDirectory, @"core\python", "cpythoncustomclass_returnManyInstances.dyn");
+            ViewModel.OpenCommand.Execute(examplePath);
+
+            var classdef = ViewModel.Model.CurrentWorkspace.Nodes.First(x => x.Name == "classdef");
+
+            ViewModel.HomeSpace.Run();
+            var handles = classdef.CachedValue.GetElements().Select(x=>x.Data).Cast<DynamoCPythonHandle>().ToList<DynamoCPythonHandle>();
+            var firstMemLoc = handles.First().PythonObjectID;
+            Assert.IsTrue(handles.All(x => x.PythonObjectID == firstMemLoc));
```

Comment:
So does this mean that all handles have the same ID?

Reply:
It means that all handles that point to the same pyObj have the same id.